libp2p/NodeTable.cpp

```diff
@@ -0,0 +1,466 @@
+	Public nodeid(dev::recover(*(Signature const*)sigBytes.data(), sha3(rlpBytes)));
+	if (!nodeid)
+	{
+		clog(NodeTableMessageSummary) << "Invalid Message signature from " << _from.address().to_string() << ":" << _from.port();
+		return;
+	}
+	noteNode(nodeid, _from);
+	
+	try {
+		switch (itemCount) {
+			case 1:
+			{
+//				clog(NodeTableMessageSummary) << "Received Pong from " << _from.address().to_string() << ":" << _from.port();
+				Pong in = Pong::fromBytesConstRef(_from, rlpBytes);
+				
+				// whenever a pong is received, first check if it's in m_evictions
+				
+				break;
+			}
+				
+			case 2:
+				if (rlp[0].isList())
+				{
+					Neighbors in = Neighbors::fromBytesConstRef(_from, rlpBytes);
+//					clog(NodeTableMessageSummary) << "Received " << in.nodes.size() << " Neighbors from " << _from.address().to_string() << ":" << _from.port();
+					for (auto n: in.nodes)
+						noteNode(n.node, bi::udp::endpoint(bi::address::from_string(n.ipAddress), n.port));
+				}
+				else
+				{
+//					clog(NodeTableMessageSummary) << "Received FindNode from " << _from.address().to_string() << ":" << _from.port();
+					FindNode in = FindNode::fromBytesConstRef(_from, rlpBytes);
+					
+					std::vector<std::shared_ptr<NodeTable::NodeEntry>> nearest = findNearest(in.target);
+					static unsigned const nlimit = (m_socketPtr->maxDatagramSize - 11) / 86;
+					for (unsigned offset = 0; offset < nearest.size(); offset += nlimit)
+					{
+						Neighbors out(_from, nearest, offset, nlimit);
+						out.sign(m_secret);
+						m_socketPtr->send(out);
+					}
+				}
+				break;
+				
+			case 3:
+			{
+//				clog(NodeTableMessageSummary) << "Received PingNode from " << _from.address().to_string() << ":" << _from.port();
+				PingNode in = PingNode::fromBytesConstRef(_from, rlpBytes);
+				
+				Pong p(_from);
+				p.replyTo = sha3(rlpBytes);
+				p.sign(m_secret);
+				m_socketPtr->send(p);
+				break;
+			}
+				
+			default:
+				clog(NodeTableMessageSummary) << "Invalid Message received from " << _from.address().to_string() << ":" << _from.port();
+				return;
+		}
+	}
+	catch (...)
+	{
+		clog(NodeTableWarn) << "Exception processing message from " << _from.address().to_string() << ":" << _from.port();
+	}
+}
+	
+void NodeTable::doCheckEvictions(boost::system::error_code const& _ec)
+{
+	if (_ec || !m_socketPtr->isOpen())
+		return;
+
+	auto self(shared_from_this());
+	m_evictionCheckTimer.expires_from_now(c_evictionCheckInterval);
+	m_evictionCheckTimer.async_wait([this, self](boost::system::error_code const& _ec)
+	{
+		if (_ec)
+			return;
+		
+		bool evictionsRemain = false;
+		std::list<shared_ptr<NodeEntry>> drop;
+		{
+			Guard le(x_evictions);
+			Guard ln(x_nodes);
+			for (auto& e: m_evictions)
+				if (chrono::steady_clock::now() - e.first.second > c_reqTimeout)
+					if (auto n = m_nodes[e.second])
+						drop.push_back(n);
+			evictionsRemain = m_evictions.size() - drop.size() > 0;
+		}
+		
+		drop.unique();
```

Comment:
drop is a list<shared_ptr<NodeEntry>>, see http://en.cppreference.com/w/cpp/container/list/unique